### DISCUSSION TITLE
Rename the various phases of the page lifecycle

### DISCUSSION THREAD

**Discussion by antocuni** [pyscript/pyscript#1238] [Proposals]
This is the follow-up of a discussion which I had in real life with @JeffersGlass @fpliger @ntoll and @philippjfr. We had a great discussion but we didn't come up with anything concrete, so let me try to revive it.

The various phases of the lifecycle of a page are currently summarized by this comment in `main.ts`:
https://github.com/pyscript/pyscript/blob/43d598d9519b7fd5e9b19803c02f6898bddfaf2e/pyscriptjs/src/main.ts#L27-L59

Moreover, during the execution we call certain hooks at specific points in time, to allow plugins to customize the behavior. So far, we introduced only the hooks which are needed for some of the (internal) plugins that we have. The various hooks are summarized here:
https://github.com/pyscript/pyscript/blob/43d598d9519b7fd5e9b19803c02f6898bddfaf2e/pyscriptjs/src/plugin.ts#L10-L73

I think that the current codebase has two problems:
1. the name of the hooks/phases don't have a clear mapping to the methods of `PyScriptApp` . For example, we have an `afterSetup` hook which is called by `afterInterpreterLoad`. The code would be much clearer if we had e.g. a `PyScript.setup()` method which calls the hooks `beforeSetup()` and `afterSetup()`. This is only internal-facing (since the end users won't see it) but it's good for the health of the codebase.

2. We should probably rename the external-facing hooks so that they are clearer to understand and more obvious.

Let's concentrate on (2) for now.

---

### Lifecycle phases

From a very high level point of view, this is what a pyscript page currently does :
1. preliminary initialization such as finding/parsing the `<py-config>` tag
2. call the `configure` hook
3. call the `beforeLaunch` hook
4. download pyodide (`PyScriptApp.loadInterpreter`)
5. start pyodide (`PyScriptApp.afterInterpreterLoad`)
6. install the relevant packages (`PyScriptApp.setupVirtualEnv`)
7. fetch and install user-defined python plugins (also in `PyScriptApp.setupVirtualEnv`)
8. fetch the files defined by `config.fetch` (still `PyScriptApp.setupVirtualEnv`)
9. call the `afterSetup` hook
10. execute the `<py-script>` tags
11. create custom elements and init event handlers
12. call the `afterStartup` hook

My idea is to identify and give good names to the relevant phases and document them (possibly using a nice diagram): this alone should help a lot the users to understand what's going on.
Then, for each phase we create a pair of `before/after` hooks, so that it will be obvious in which order they are executed. Note that by doing that we are duplicating some hooks (e.g., `afterPhase1` and `beforePhase2` are executed basically at the same moment), but it also makes it easier to define a reasonable execution order between plugins: e.g., `beforePhase2` will be executed after the main logic AND all the plugins have finished the phase1.

Now, the hard part is to find the names. The current `launch/execute/startup` names are confusing, because one could easily argue that "startup" comes before "launch", "execute" comes after "startup", etc. etc.
Also, we should try to avoid using names which have a different meaning in the browser (e.g. `load`).

Let me propose some names to start the discussion. For each phase I propose a name and some alternatives. I really dislike some of the alternatives but I put them anyway with the hope of sparkling some creativity: feel free to propose more, that's the point of the discussion :):

1. `configure` phase: in this phase `beforeConfigure` receives a `config` object and can still modify it. Then the config is frozen and in `afterConfigure` each plugin can validate its own config

2. `preload` phase: this is when we download pyodide (and possibly other stuff). **Alternatives**: `download`, `preinit`, `prefetch`, `downloadDependencies`.

3. `initInterpreter` phase: this is when pyodide starts. **Alternatives**: `startInterpreter`, `start`, `launch`, `init`.

4. `setupEnvironment` phase: this is when we download/install packages and fetch user files. **Alternatives**: `setupVirtualEnv`, `installPackages`, `createEnvironment`, `environment`, `setup`.

5. `execute` phase: this is where we execute the code in `<py-script>` tags. **Alternatives**: `exec`, `pyscript`, `script`, `main`, `userScripts`

6. `DOMSetup` phase: this is where we create the remaining custom elements (`py-repl`, etc.) and init the event handlers. I don't really like this name but cannot think of anything better right now.

7. `idle` phase: the startup is complete, the splashscreen is removed and now the user can interact with the page: this phase will never end. **Alternatives**: `main`, `user`, `interactive`, `interaction`, `page`. Note that if we are consistent with naming, the hook will be called `beforeIdle`. We could make an exception and call it e.g. `afterStartup` or similar to indicate that that the *whole* startup sequence has ended.

**Comment by JeffersGlass** [comment]
Thank you for laying this all out! In general I think the names you have are quite good, and I very much like the idea of every phase having a `before[Phase]` and an `after[Phase]` method, and those being the names for the primary external hooks.

General thought: Since the phase names are going to be the names of a stretch of time, instead of the actual methods themselves, I wonder if it makes sense for the phase names to be nouns instead of verbs, i.e. we would be able to say _"during the 'execution' phase"_ instead of _"during the 'execute' phase"_.

For phase 1,  the only thing that's happening _within_ that phase is that the configuration is being frozen, right? (Whether we actually "freeze" that in a JS sense, or just instruct users that that is correct behavior.) What if that phase's methods were `beforeConfigFreeze()` and `afterConfigFreeze()`? I suppose it could also be `modifyConfig()` and `validateConfig()`, but making them explicitly about what they are before and after I think is better.

As for phase 6... I'm just musing about some kind of common description that encompasses both custom elements like py-repl and event handlers, as a way of coming up with names... maybe they're "PyScript Utility Features" that get setup in a `Utility Setup` phase? Or they both involve HTML elements and attributes... `Element Setup` phase? Don't love that. I think I agree with you that DOMSetup doesn't feel quite right...

So, all together, what I'm suggesting:

| # | Phase Name | Pre-method | Post-method|
|-|-|-|-|
| 1 | Configuration Phase | `beforeConfigFreeze()` | `afterConfigFreeze()` |
| 2 | Preload Phase | `beforePreload()` | `afterPreload()`|
| 3 | Interpreter Initialization Phase | `beforeInterpreterInit()` | `afterInterpreterInit()`|
| 4 | Environment Setup Phase | `beforeEnvironmentSetup()` | `afterEnvironmentSetup()`|
| 5 | Execution Phase | `beforeExecution()` | `afterExecution()`|
| 6 | Utility Setup Phase | `beforeUtilitySetup()` | `afterUtilitySetup()`|
| 7 | Idle Phase | `beforeIdle()`| |

**Reply by antocuni** [reply]
>  I wonder if it makes sense for the phase names to be nouns instead of verbs, i.e. we would be able to say "during the 'execution' phase" instead of "during the 'execute' phase".

The original idea of using verbs was to match very precisely the name of the corresponding methods of `PyScriptApp`: e.g., if you have `PyScriptApp.execute()` you have `beforeExecute`.
But I agree that from the end user point of view it might be better/clearer to use nouns, so +1 for this.


> What if that phase's methods were `beforeConfigFreeze()` and `afterConfigFreeze()`

+0 for this. I don't dislike it, although I'm vaguely not convinced (but I can't express exactly why 😅)


> As for phase 6... I'm just musing about some kind of common description that encompasses both custom elements like py-repl and event handlers, as a way of coming up with names... maybe they're "PyScript Utility Features" that get setup in a `Utility Setup` phase? Or they both involve HTML elements and attributes... `Element Setup` phase? Don't love that. I think I agree with you that DOMSetup doesn't feel quite right...

I fear this is going to be the hardest name of all. `UtilitySetup` is very obscure and it sounds a bit like a "catch-all" name. I would like something better but again, I don't have many ideas.

The fact that we can't find a proper name might indicate that it's not a logical distinct phase. It's also possible that we don't need it at all: if we move all the builtin logic of py-repl and even handlers into plugins, they they could just use the `afterExecution()` hook. I think that the question is: do we need this level of granularity?

**Comment by JeffersGlass** [comment]
In thinking about how we might include a nice diagram of the phases, in a way that would make it easy for anyone to edit without specific software, I've been playing with [GitHub's built-in Mermaid Diagram renderer](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams).

Of course the names aren't final and the layout isn't amazing yet, but if you look at the source of this comment, you can see that the diagram is specified in a markdown-like syntax right in the text (not as an image), which I think could be nice.

```mermaid
flowchart TB
    Configuration:::phaseName
    subgraph Configuration
        A{{"beforeConfigFreeze()"}}:::method -->
        B{{"afterConfigFreeze()"}}:::method
    end
    Preload:::phaseName
    subgraph Preload
        C{{"beforePreload()"}}:::method -->
        D("Download Pyodide"):::action -->
        E{{"afterPreload()"}}:::method
    end
    InitInterpreter:::phaseName
    subgraph InitInterpreter ["Initialize Interpreter"]
        F{{"beforeInterpreterInit()"}}:::method -->
        G("Initialize Pyodide"):::action -->
        H{{"beforeInterpreterInit()"}}:::method
    end
    Environment:::phaseName
    subgraph Environment ["Environment Setup"]
        I{{"beforeEnvironmentSetup()"}}:::method -->
        J1("Download/install user packages"):::action -->
        J2("Fetch user files"):::action -->
        K{{"AfterEnvironmentSetup()"}}:::method
    end
    Execution:::phaseName
    subgraph Execution
        L{{"beforeExecution()"}}:::method -->
        M("Execute py-script tags"):::action -->
        N{{"afterExection()"}}:::method
    end
    Utility:::phaseName
    subgraph Utility ["Utility Setup"]
        O{{"beforeUtilitySetup()"}}:::method -->
        P1("Setup py-repl elements"):::action -->
        P2("Initialize py-[event] handlers"):::action -->
        Q{{"afterUtilitySetup()"}}:::method
    end
    Idle:::phaseName
    subgraph Idle
        IdleMethod{{"beforeIdle()"}}:::method
    end

    Configuration --> Preload
    Preload --> InitInterpreter
    InitInterpreter --> Environment
    Environment --> Execution
    Execution --> Utility
    Utility --> Idle

%%classDef method fill: #f00;
classDef default fill:#000,stroke:#333,stroke-width:2px;
classDef method fill:#131,stroke:#333,stroke-width:2px;
classDef action fill:#338,stroke:#333,stroke-width:2px;

classDef phaseName color:#ddd, font-size: 1.5rem;




```

**Reply by antocuni** [reply]
wow, I love the idea! And yes, being able to edit it without any specific software is a big plus.
There is some problem with colors on my machine though: the text is black on dark background, which makes it impossible to read:
![image](https://user-images.githubusercontent.com/344507/227153875-3226225f-100c-4895-be7f-0e9df342a304.png)

**Comment by fpliger** [comment]
Finally catching up on all the pending discussions... +1 on the diagram, it's really super useful.

I'm +1 with you both that "Utility Setup" is less than good and, if I had to choose, I'd be +1 on `DOMSetup`, even if it feels like it's leaking implementation details. Anyway, I can't think of a better name right now though :/

**Comment by bugzpodder** [comment]
I did some console logs and I noticed that `configure` and `beforeLaunch` isn't getting called for custom js plugins.  Most likely because the custom plugins needed to be loaded async since only a string url is supported?

**Reply by JeffersGlass** [reply]
Hmmm interesting - this is probably worth submitting an issue for if you're up for it, as it's a good catch, and slightly off topic for the naming discussion.

**Reply by bugzpodder** [reply]
I think the current lifecycle is:
1. loadConfig
2. plugin.configure, plugin.beforeLaunch
3. loadInterpreter
    * afterInterpreterLoad
        * setupVirtualEnv
            * fetchUserPlugins

And this was posted initially at the top of this thread here as well: https://github.com/pyscript/pyscript/discussions/1238#discussion-4911514

So userPlugin doesn't get configure/beforeLaunch events.

**Reply by JeffersGlass** [reply]
Ah gotcha, so similar to #1219 then? Agreed, moving the fetching of JS plugins earlier would be great.

The reason they're fetched so late (after the Python interpreter is fully loaded) I think is just an accident of history. We used to only have Python plugins, so it made sense to fetch and initialize them after the Python interpreter was loaded. Now we have JS plugins as well, but all are still being fetched at that late point in the lifecycle. So no real reason they couldn't be moved earlier.
